### ISSUE TITLE
Vectorize API

### ISSUE THREAD

**Issue by alexksikes** [elastic/elasticsearch#10823]
The vectorize API operates on top of the Term Vectors API and on `_source` in order to return a document-term matrix according to some user given specification. In a document-term (or term-document matrix), the rows correspond to documents in an index and the columns correspond to terms, or more precisely to some numerical value associated with each term such as tf or tf-idf.

## Defining the Doc-Term Matrix

First a vectorizer has to be registered. A vectorizer defines the shape and entries of the desired outputted matrix.

```
PUT /index/.vectorizer/my_vectorizer
{
    "vectorizer": [
        {
            "field": "text",
            "span": [ ... list of terms ...],
            "value": "term_freq"
        },
        {
            "field": "field_numeric_1",
            "span": 1
        },
        {
            "field" : "field_numeric_2",
            "span": 1,
            "script": "if _value > 0.5 then 1 else 0"
        },
        {
            "field": "field_numeric_3",
            "span": 5
        },
        {
            "field": "label",
            "span": 1,
            "script": "if _value == "yes" then 1 else 0"
        }
    ]
}
```

In this example, the first entry defines a first list of columns whose values are the term frequencies of the terms given in `span` in this order. The next column is simply defined as the value of the numerical field named `field_numeric_1`. For the next column, we binarize the numerical value at the field using a script. Next we assign 5 columns for the first 5 values in the multi-valued numerical field `field_numeric_3`. Last but not least, the last column is binarized as well and could be used as the target label for learning.

More precisely:

* `field`: Name of the field in the index to which a mapping from field values to real values should be applied.

* `span`: For numerical fields, specifies a number of columns to be assigned in the resulting vector. For string fields, specifies a list of terms, each term occupying a column in the vector if present in the document.

* `value`: For numerical fields, defaults to the value at the field. For string fields, defaults to term frequency. For the later case other values are possible such as document frequency, or payload.

* `script`: Allows for any field value transformation such as thresholding, categorizing, etc ...

## Retrieving Vectors

We can now use the vectorize API:

```
GET /index/type/id/_vectorize&vectorizer=my_vectorizer

Response:

{
    "shape": [1, 9],
    "vector": [
        {"3": 5, "5": 2, "6": 0.55, 7": 1, "8": 0}
    ]
}
```

The matrix returned is in a sparse format, together with its dimension. This vector can then be loaded in memory in your favorite statistical environments, or fed to a machine learning package. The parameters are exactly the same as the term vectors API with the additional `vectorizer` parameter. A vectorizer could also be provided inline, meaning as part of the request.

```
GET /index/type/id/_vectorize
{
    "vectorizer": [
        {
            "field": "text",
            "span": [ ... list of terms ...],
            "value": "term_freq"
        },
        {
            "field": "label",
            "span": 1,
            "script": "if _value == "yes" then 1 else 0"
        }
    ]
}
```

In order to retrieve more than one vector, use the `_mvectorize` API:

```
GET /index/type/_mvectorize
{
    "docs": [
    {
        "_index": "index",
        "_type": "type",
        "_id": "id",
        "vectorizer": my_vectorizer
    },
    {
        "_index": "index",
        "_type": "type",
        "_id": "id2",
        "vectorizer": my_vectorizer
    }
   ]
}

Response:

{
    "shape": [2, 9],
    "vector": [
        {"3": 5, "5": 2, "6": 0.55, 7": 1, "8": 0},
        {"0": 2, "3": 1, "6": 0.21, 7": 1, "8": 1}
    ]
}
```

Here different vectorizers could have been chosen. In this case, the vectors are stacked up, columns over columns, with the largest vector defining the dimension of the matrix.

## Generating Datasets

Finally, we can obtain a dataset by using a new search type called `vectorize` together with a scroll:

```
GET /index/type/_search?search_type=vectorize&scroll=1m
{
    "query": { "match_all": {}},
    "size":  1000,

    "from": 0,
    "step": 1,
    "sample": "10%",

    "vectorizer": "my_vectorizer",
    "slice": start:stop:step
}

After submitting the scroll request, we get the response:

{
    "shape": [1000, 9],
    "vector: [
        {"3": 5, "5": 2, "6": 0.55, 7": 1, "8": 0},
        {"0": 2, "3": 1, "6": 0.21, 7": 1, "8": 1},
        {"3": 5, "3": 5, "6": 0.45, 7": 0, "8": 0},
        {"3": 5, "5": 3, "6": 0.56, 7": 0, "8": 1},
        ...
    ]
}
```

Here `slice` allows us to only select some columns in this matrix. The `from`, `step` or `sample` retain documents starting after having processed a given number of documents and at every chosen step, or with x% chances. These options are useful in order to generate datasets such as a training and test set.

**Comment by polyfractal** [comment]
I like it!  I can definitely see this being used with plug-n-play tools like WEKA or scikit-learn.  Some questions / thoughts:

- Does everything just throw an exception if term vectors are not enabled?  Is this a hard requirement or just a way to improve speed?

- Why is `_source` required?

- How are missing values handled?  Is `null` or `NaN` encoded in the matrix if the document is missing a particular field?  I know some ML packages are particular about how they handle missing fields, perhaps there could be an optional `missing_value` parameter that inserts a default if the field is missing?  Same for spans + numerics.  E.g. if you request `span: 5`, but a field only has 3 values ... the extra two will likely need to be filled with `NaN` or `null` or something?

- With `_mvectorize`, is there an option to "unstack" the matrices?  E.g. you may just want to request multiple matrices in one request, but not combine them.

- The slice syntax seems clunky, and only sorta flexible.  E.g what if you want uneven combinations such as `0-5, 7, 9, 10-12` ?  Perhaps some kind of array syntax:  `[0..5, 7, 9, 10..12]` ?

- Perhaps show the resulting "shape" after `slice`ing a matrix to make it clear it is different from the underlying vectorizer?

- Would it be useful to also add the vectorizer name to the output?  Not sure if that is something people would want.

- Not necessarily a problem, just something to think about, but `step` and `sample` could get strange if a routing scheme is used.  E.g. `sample: 10%` may exhaust some shards faster than others, because certain shards are "hot" due to routing.  This could unbalance the sampling and skew it towards the particular routed values.  Might just be a documentation issue so that users are aware sampling doesn't necessarily mean truly random sample depending on circumstances.

- I wonder if there should be a "csv" or "tab" format option, so you can get a flat matrix rather than JSON?  Might make it easier to interface with other packages?  Not sure...it's very un-ES to emit non-JSON :)

**Comment by alexksikes** [comment]
Thank you for reading through this. 

> Does everything just throw an exception if term vectors are not enabled? Is this a hard requirement or just a way to improve speed?

Term vectors are generated on the fly if they are not stored in the index, so this should not be an issue, only that vectorizing term vectors would be slower in this case.

> Why is _source required?

This is in order to retrieve numerical values which the TVs API does not return. But we don't necessarily need _source to be enabled, as long as the values are stored.

> How are missing values handled? Is null or NaN encoded in the matrix if the document is missing a particular field? I know some ML packages are particular about how they handle missing fields, perhaps there could be an optional missing_value parameter that inserts a default if the field is missing? Same for spans + numerics. E.g. if you request span: 5, but a field only has 3 values ... the extra two will likely need to be filled with NaN or null or something?

If the document is missing a particular field then it means that it has none of these terms, so this part of the vector would zero, meaning not shown in the sparse vector format representation. The same behavior would apply if the some value could not be fetched. We could also fill up with a default missing values, but in this case the vector would no longer be sparse, leading to potentially huge outputs.

> With _mvectorize, is there an option to "unstack" the matrices? E.g. you may just want to request multiple matrices in one request, but not combine them.

A vectorizer defines the desired outputted matrix. It is true that stacking these vectors is arbitrary, but all these (say m) vectors retrieved would have the same shape (say of length n), naturally leading to a matrix of shape m by n.

> The slice syntax seems clunky, and only sorta flexible. E.g what if you want uneven combinations such as 0-5, 7, 9, 10-12 ? Perhaps some kind of array syntax: [0..5, 7, 9, 10..12] ?

Yes that could be an option, definitely more flexible. Something which could be worth exploring is returning multiple matrices under different array syntax combinations. This in combination with `from`, `step` and `sample` would then return multiple matrices. So instead we would not have to go through different runs just in order to generate say a training and test set.

> Perhaps show the resulting "shape" after sliceing a matrix to make it clear it is different from the underlying vectorizer?

Thanks for catching that, indeed if we are slicing the matrix then the resulting shape should be different as well.

> Would it be useful to also add the vectorizer name to the output? Not sure if that is something people would want.

Yes that makes sense, but if the vectorizer is "inlined", meaning as part of the request, then that could mean returning potentially a lot of terms.

> Not necessarily a problem, just something to think about, but step and sample could get strange if a routing scheme is used. E.g. sample: 10% may exhaust some shards faster than others, because certain shards are "hot" due to routing. This could unbalance the sampling and skew it towards the particular routed values. Might just be a documentation issue so that users are aware sampling doesn't necessarily mean truly random sample depending on circumstances.

Yes it should definitely be documented, sampling would be true to the shard, not to the index.

> I wonder if there should be a "csv" or "tab" format option, so you can get a flat matrix rather than JSON? Might make it easier to interface with other packages? Not sure...it's very un-ES to emit non-JSON :)

We have plans to make this into a plugin, and yes in this case it would make a lot of sense to support different output formats.

**Comment by polyfractal** [comment]
> If the document is missing a particular field then it means that it has none of these terms, so this part of the vector would zero, meaning not shown in the sparse vector format representation. 

Oh, I see... the key in the vector is numbered to represent the column...missed that on the first pass.  Gotcha :)

> A vectorizer defines the desired outputted matrix. It is true that stacking these vectors is arbitrary, but all these (say m) vectors retrieved would have the same shape (say of length n), naturally leading to a matrix of shape m by n.

Well, my question was more if you wanted to `_mvectorize` two different vectorizers:  "my_vectorizer" with length `n`, and "my_other_vectorizer" with length `j`.  Basically like all the other multi-APIs, like _mget or _mpercolate, where you are just batching stuff together.

Something like this:
```
GET /index/type/_mvectorize
{
    "docs": [
    {
        "vectorizer": "my_vectorizer"
    },
    {
        "vectorizer": "my_other_vectorizer"
    }
   ]
}
```

**Comment by alexksikes** [comment]
> Well, my question was more if you wanted to _mvectorize two different vectorizers: "my_vectorizer" with length n, and "my_other_vectorizer" with length j. Basically like all the other multi-APIs, like _mget or _mpercolate, where you are just batching stuff together.

I had just updated the comment as you were writing this :) Yes basically it would make sense to only stack matrices according to the vectorizer that they have in common.

**Comment by clintongormley** [comment]
> Well, my question was more if you wanted to _mvectorize two different vectorizers: "my_vectorizer" with length n, and "my_other_vectorizer" with length j. Basically like all the other multi-APIs, like _mget or _mpercolate, where you are just batching stuff together.

+1 to named vectors, just like we do with aggs, suggesters, etc

> `from`, `step` and `sample``

The `from` parameter really has no meaning when scanning is used.  `step` and `sample` are probably best implemented using the `function_score` query, with a random function or a modulus on the ID or something, instead of adding new top level parameters.

**Comment by alexksikes** [comment]
I rebased the vectorizer branch on master and added some tests. I'll create a new branch for the actual plugin but let's have this at least for reference.

https://github.com/alexksikes/elasticsearch/commits/feature/tvs-vectorize

**Comment by alexksikes** [comment]
For reference, this is now a plugin. Please keep in mind this is still very alpha.

https://github.com/alexksikes/elasticsearch-vectorize/

**Comment by diugalde** [comment]
Hello, Is there any current work on this feature?

**Comment by s1monw** [comment]
no this is not being worked on and won't happen any time soon... closing for now

**Comment by pkphlam** [comment]
Is there a reason why there is no work planned on this any time soon? Given the focus and rage about Machine Learning, this seems like an excellent feature that could potentially make ES the de facto datastore for a lot of machine learning/statistical work, especially with respect to text analytics. Currently, trying to replicate something like this requires a ton of work on both client and ingest sides with very little efficiency. There is really very little other software stacks out there that can combine ES's great query capabilities with a machine learning pipeline in an efficient way and this could be a gamechanger. Definitely a +1 for this feature.
